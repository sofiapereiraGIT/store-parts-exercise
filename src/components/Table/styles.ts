import styled from "styled-components";
import { Table, TableHead, TableRow, TableCell } from "@mui/material";

export const TableStyles = {
  NoContentWrapper: styled.div`
    width: 100%;
    display: flex;
    justify-content: center;
    margin: 0.5em;
  `,
  Table: styled(Table)`
    && {
      min-width: 300px;
    }
    margin-top: 0.5em;
  `,
  TableHead: styled(TableHead)`
    background-color: black;
  `,
  HeaderTableCell: styled(TableCell)`
    && {
      color: white;
      font-size: 1em;
      font-weight: bold;
    }
  `,
  BodyTableRow: styled(TableRow)`
    &:nth-of-type(odd) {
      background-color: #f1f1f1;
    }
    :hover {
      cursor: pointer;
      background-color: lightgrey;
    }
  `,
};
