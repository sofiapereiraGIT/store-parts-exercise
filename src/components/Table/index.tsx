import { useHistory } from "react-router-dom";
import _ from "lodash";
import {
  CircularProgress,
  TableBody,
  TableCell,
  TableRow,
} from "@mui/material";
import { Item } from "../../models/item";
import { TableStyles as S } from "./styles";

export default function Table(props: { loading: boolean; items: Item[] }) {
  const history = useHistory();

  return (
    <>
      <S.Table aria-label="customized table">
        <S.TableHead>
          <TableRow>
            <S.HeaderTableCell>Name</S.HeaderTableCell>
            <S.HeaderTableCell>Type</S.HeaderTableCell>
            <S.HeaderTableCell align="right">Price</S.HeaderTableCell>
          </TableRow>
        </S.TableHead>

        <TableBody>
          {props.items.map((item: Item, index) => (
            <S.BodyTableRow
              key={`${item.name}_${item.type}_${item.price}_${index}`}
              data-testid={`bodyTableRow_${index}`}
              onClick={() =>
                history.push({
                  pathname: "/item",
                  state: item,
                })
              }
            >
              <TableCell>{item.name}</TableCell>
              <TableCell>{item.type}</TableCell>
              <TableCell align="right">{item.price}</TableCell>
            </S.BodyTableRow>
          ))}
        </TableBody>
      </S.Table>

      <S.NoContentWrapper>
        {props.loading ? (
          <CircularProgress data-testid="loadingCircle" />
        ) : (
          _.isEmpty(props.items) && (
            <span data-testid="noDataMessage">No data available</span>
          )
        )}
      </S.NoContentWrapper>
    </>
  );
}
