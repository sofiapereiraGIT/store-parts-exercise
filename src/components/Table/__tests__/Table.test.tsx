import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import Table from "../index";

describe("Table Tests", () => {
  it("Renders properly when items are provided", () => {
    const renderedComponent = renderer.create(
      <Table
        loading={false}
        items={[
          {
            name: "itemName 1",
            type: "itemType",
            price: "11$",
          },
          {
            name: "itemName 2",
            type: "itemType",
            price: "12$",
          },
          {
            name: "itemName 3",
            type: "itemType",
            price: "13$",
          },
        ]}
      />
    );
    expect(renderedComponent.toJSON()).toMatchSnapshot();
  });

  it("Shows loading", () => {
    const { getByTestId } = render(<Table loading={true} items={[]} />);
    const loadingCircle = getByTestId("loadingCircle");
    expect(loadingCircle).toBeTruthy();
  });

  it("Shows 'no data' message", () => {
    const { getByTestId } = render(<Table loading={false} items={[]} />);
    const noDataMessage = getByTestId("noDataMessage");
    expect(noDataMessage).toBeTruthy();
  });

  it("Redirects properly to '/item' when a row is clicked", () => {
    const history = createMemoryHistory();
    const pushSpy = jest.spyOn(history, "push");
    const { getByTestId } = render(
      <Router history={history}>
        <Table
          loading={false}
          items={[
            {
              name: "itemName 1",
              type: "itemType",
              price: "11$",
            },
            {
              name: "itemName 2",
              type: "itemType",
              price: "12$",
            },
            {
              name: "itemName 3",
              type: "itemType",
              price: "13$",
            },
          ]}
        />
      </Router>
    );

    const bodyTableRow = getByTestId("bodyTableRow_0");
    fireEvent.click(bodyTableRow);
    expect(pushSpy).toHaveBeenCalledWith({
      pathname: "/item",
      state: {
        name: "itemName 1",
        type: "itemType",
        price: "11$",
      },
    });
  });
});
