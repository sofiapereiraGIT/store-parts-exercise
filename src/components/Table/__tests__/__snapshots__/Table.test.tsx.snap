// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Tests Renders properly when items are provided 1`] = `
Array [
  <table
    aria-label="customized table"
    className="MuiTable-root sc-gsnTZi dTKQAJ css-rqglhn-MuiTable-root"
    role={null}
  >
    <thead
      className="MuiTableHead-root sc-dkzDqf lnvFuf css-15wwp11-MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium sc-hKMtZM dmwtRg css-1ygcj2i-MuiTableCell-root"
          scope="col"
        >
          Name
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium sc-hKMtZM dmwtRg css-1ygcj2i-MuiTableCell-root"
          scope="col"
        >
          Type
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium sc-hKMtZM dmwtRg css-1azl6jz-MuiTableCell-root"
          scope="col"
        >
          Price
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root sc-eCYdqJ eKANlz css-1q1u3t4-MuiTableRow-root"
        data-testid="bodyTableRow_0"
        onClick={[Function]}
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
        >
          itemName 1
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
        >
          itemType
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
        >
          11$
        </td>
      </tr>
      <tr
        className="MuiTableRow-root sc-eCYdqJ eKANlz css-1q1u3t4-MuiTableRow-root"
        data-testid="bodyTableRow_1"
        onClick={[Function]}
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
        >
          itemName 2
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
        >
          itemType
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
        >
          12$
        </td>
      </tr>
      <tr
        className="MuiTableRow-root sc-eCYdqJ eKANlz css-1q1u3t4-MuiTableRow-root"
        data-testid="bodyTableRow_2"
        onClick={[Function]}
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
        >
          itemName 3
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
        >
          itemType
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
        >
          13$
        </td>
      </tr>
    </tbody>
  </table>,
  <div
    className="sc-bczRLJ gFctAL"
  />,
]
`;
